test/integration/fake_upstream.h

```diff
@@ -240,8 +241,13 @@ class FakeHttpConnection : public Http::ServerConnectionCallbacks, public FakeCo
     ReadFilter(FakeHttpConnection& parent) : parent_(parent) {}
 
     // Network::ReadFilter
-    Network::FilterStatus onData(Buffer::Instance& data, bool) override {
-      parent_.codec_->dispatch(data);
+    Network::FilterStatus onData(Buffer::Instance& data, bool end_stream) override {
+      try {
+        parent_.codec_->dispatch(data);
+      } catch (Http::CodecProtocolException& e) {
+        ASSERT(end_stream);
```

Comment:
This is pretty sketchy. Why this change?

Reply:
Since the FakeUpstream doesn't fake well compare to real [codec_client.onData](https://github.com/envoyproxy/envoy/blob/master/source/common/http/codec_client.cc#L99) called by its [ReadFilter](https://github.com/envoyproxy/envoy/blob/master/source/common/http/codec_client.h#L138).

The http_parser IS half-close aware and if you feed it with lengh=0 data, it will throw INVALID_EOF which is expected, so the FakeUpstream should just handle it correctly.